File: ContigousArray.cs
//Time Complexity: O(n) as we have to iterate through the array
//Space Complexity: O(n)
public class Solution {
     public int FindMaxLength(int[] nums)
        {
            if (nums == null || nums.Length == 0)
            {
                return 0;
            }
            Dictionary<int, int> map = new Dictionary<int, int>();
            int cnt = 0; int max = 0;
            int start = -1, end = -1;
            map.Add(0, -1);
            for (int i = 0; i < nums.Length; i++)
            {
                if (nums[i] == 0)
                {
                    cnt = cnt - 1;
                }
                else
                {
                    cnt = cnt + 1;
                }
                if (!map.ContainsKey(cnt))
                {
                    map.Add(cnt, i);
                }
                else
                {
                    if(max < i - map[cnt])
                    {
                        max = i - map[cnt];
                        start = map[cnt];
                        end = i;
                    }
                    
                }
            }
         if(start == -1) start = 0;
         if(max != 0) Console.WriteLine(start +"-" + end);
          return max;
        }
}

File: Contigous_Array.py
#Time Complexity - O(n)
#Space Complexity - O(n)
#Successfully compiled and run on Leetcode


class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        currsum = 0
        maxi = 0

        hm = {}
        hm[0] = -1 # init of hashmap

        for i in range(len(nums)):
            if (nums[i] == 0): #if num is 1 = init with -1 or num is 0 = init with 1
                currsum += 1
            else:
                currsum -= 1

            if currsum in hm: # check if running sum exists
                end = i
                start = hm[currsum]

                maxi = max(maxi, (end - start)) # then calcuate the difference in its first seen index and current index
                # maxi will have max length of contigous array
            else:
                hm[currsum] = i #else just add the entry into hashmap

        return maxi